api/v1alpha1/gateway_types.go

```diff
@@ -54,78 +57,129 @@ type GatewayList struct {
+	TCPProtocolType ProtocolType = "TCP"
 )
 
-// Listener defines a
+// Listener embodies the concept of a logical endpoint where a
+// Gateway can accept network connections.
+//
+// The GatewayClass may collapse compatible Listener definitions
+// into single implementation-defined acceptor.
+//
+// Listeners are considered compatible if the Protocol field is
+// "HTTP", HTTPS or "TLS", the Hostname field is specified, and the
+// Port field matches.
 type Listener struct {
-	// Name is the listener's name and should be specified as an
-	// RFC 1035 DNS_LABEL [1]:
+	// Hostname is the fully qualified domain name of a network host,
+	// as defined by [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.2.2).
+	// Note the following deviations from the "host" part of the URI as defined in the RFC:
 	//
-	// [1] https://tools.ietf.org/html/rfc1035
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
 	//
-	// Each listener of a Gateway must have a unique name. Name is used
-	// for associating a listener in Gateway status.
+	// Hostname specifies the virtual hostname for protocol
+	// types that define this concept. If the protocol is "HTTP"
+	// or "HTTPS", Hostname is must be present and must not be an
+	// empty string.
 	//
-	// Support: Core
+	// Incoming requests are matched against Hostname before processing HTTPRoute
+	// rules. For example, if the request header contains host: foo.example.com,
+	// an HTTPRoute with hostname foo.example.com will match. However, an
+	// HTTPRoute with hostname example.com or bar.example.com will not match.
+	// If Hostname is unspecified, the Gateway routes all traffic based on
+	// the specified rules.
 	//
-	// +required
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
-	// Address requested for this listener. This is optional and behavior
-	// can depend on GatewayClass. If a value is set in the spec and
-	// the request address is invalid, the GatewayClass MUST indicate
-	// this in the associated entry in GatewayStatus.Listeners.
+	// Hostname must be supplied required for "HTTP" and "HTTPS" protocols, and may
+	// be supplied for "TLS" protocol.
 	//
-	// Support:
+	// Support: Core
 	//
 	// +optional
-	Address *ListenerAddress `json:"address,omitempty" protobuf:"bytes,2,opt,name=address"`
-	// Port is a list of ports associated with the Address.
+	// +kubebuilder:validation:MinLength=1
+	Hostname string `json:"hostname,omitempty" protobuf:"bytes,1,opt,name=hostname"`
+
+	// Port is the network port. Multiple listeners may use the
+	// same port, subject to the Listener compatibility rules.
 	//
-	// Support:
-	// +optional
-	Port *int32 `json:"port,omitempty" protobuf:"varint,3,opt,name=port"`
-	// Protocol to use.
+	// Support: Core
 	//
-	// Support:
-	// +optional
-	Protocol *string `json:"protocol,omitempty" protobuf:"bytes,4,opt,name=protocol"`
+	// +required
+	Port int32 `json:"port,omitempty" protobuf:"varint,3,opt,name=port"`
```

Comment:
Under this model it seems like it would be helpful to have multiple ports/protocols per listener. I know we'd already discussed that as an option, but now that routes are per listener, it seems to be even more necessary.

Reply:
There's no need for that, and I'd argue that it doesn't work. You can't have multiple processes accepting on the same socket unless you know the application protocol and that protocol has some method of discrimination. This means that you are, by definition, doing something protocol-specific, so trying to make that generic is a false abstraction that obscures rather than clarifies.

In this model, you can have a listener per port, and the rules for collapsing listeners onto the same underlying network port are clear. If you need more ports, then add more listeners with the same selector.

A further aspect that I didn't draw out in this PR is that the protocol and the route selector fields are tightly coupled and there have to be rules about how to handle this. Tying them together in the same struct makes it easier to elaborate the rules for handling that.

Reply:
Yep, as I've thought about this more it really does make sense to keep listeners unique per port/protocol combination.